pkg/apis/volumesnapshot/v1alpha1/types.go

```diff
@@ -0,0 +1,338 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+
+	core_v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+const (
+	// VolumeSnapshotDataResourcePlural is "volumesnapshotdatas"
+	VolumeSnapshotDataResourcePlural = "volumesnapshotdatas"
```

Comment:
Do you have a suggestion for this?  "data" is a plural itself.  How do we differentiate between singular and plural if they are the same?

Reply:
Looks like "volumesnapshotdatas" will be generated automatically for clientset when generating code for the CRD. So unless if we rename VolumeSnapshotData to something else, it will be "volumesnapshotdatas".

Reply:
> Looks like "volumesnapshotdatas" will be generated automatically for clientset when generating code for the CRD. So unless if we rename VolumeSnapshotData to something else, it will be "volumesnapshotdatas".

That's correct, and is a gap in the client generator. This has also been reported by others (see https://kubernetes.slack.com/archives/C0EG7JC6T/p1528416072000218). However, we shouldn't accept an incorrect API because of that... it would be better to participate in improving the generator to allow specifying the plural.